### PR TITLE
Add missing validation for replication

### PR DESCRIPTION
**Pull request by poornas**
## Motivation and Context
While adding a replication target, currently a validation is missed to check if target bucket supports versioning, 

## How to test this PR?
Use mc replication [PR](https://github.com/minio/mc/pull/3272)

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

File: cmd/bucket-replication.go [hunk 1]
```diff
@@ -84,6 +84,10 @@ func (sys *BucketReplicationSys) SetTarget(ctx context.Context, bucket string, t
 	if !ok {
 		return BucketReplicationDestinationNotFound{Bucket: tgt.TargetBucket}
 	}
+	vcfg, err := clnt.GetBucketVersioning(ctx, tgt.TargetBucket)
+	if err != nil || vcfg.Status != "Enabled" {
```
Comment (harshavardhana) [inline cmd/bucket-replication.go]:
```suggestion
	if err != nil || vcfg.Status != string(versioning.Enabled) {
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Please also add check to disable all APIs which are not supported on FS mode @poornas - other than GetBucket* calls we should disable all PutBucket* operations.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#dea7dece-0b54-40f6-a056-830257fbd4c5) |
| mint-large-bucket.sh | :x: [more...](#2e7c9093-537f-4e96-8c28-a38955cc3f83) |
| mint-dist-xl.sh | :x: [more...](#1d4eaed4-efe6-4bb6-8d29-89eb4b17846a) |
| mint-zoned.sh | :x: [more...](#0f484526-f6f0-4d74-a680-f54ced17dcd4) |
<a name="0f484526-f6f0-4d74-a680-f54ced17dcd4"></a>
### 10114-7b813ab/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:31020
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bd5536032020:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 2 minutes and 12 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 24 seconds
(8/15) Running minio-dotnet tests ... done in 47 seconds
(9/15) Running minio-go tests ... FAILED in 2 minutes and 1 seconds
{
  "args": {
    "bucketName": "minio-go-test-6l02p3uzunyy41s2",
    "objectName": "5fj91jqj06lqsxcl6nilrzshhizk15",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 4764,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 1 minutes and 24 seconds
(10/15) Running minio-js tests ... done in 55 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 33 seconds
(12/15) Running s3cmd tests ... done in 35 seconds
(13/15) Running s3select tests ... done in 9 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="1d4eaed4-efe6-4bb6-8d29-89eb4b17846a"></a>
### 10114-7b813ab/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31942
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d45d66e466f6:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 2 minutes and 14 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 22 seconds
(8/15) Running minio-dotnet tests ... done in 45 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 53 seconds
{
  "args": {
    "bucketName": "minio-go-test-eyinfuywhavtwike",
    "objectName": "m5k4oq4c92bobouwmcuxnfxm4yy950",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 5486,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 1 minutes and 24 seconds
(10/15) Running minio-js tests ... done in 51 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 20 seconds
(12/15) Running s3cmd tests ... done in 20 seconds
(13/15) Running s3select tests ... done in 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="dea7dece-0b54-40f6-a056-830257fbd4c5"></a>
### 10114-7b813ab/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:30554
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ff99bf993b48:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 1 minutes and 12 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 25 seconds
(8/15) Running minio-dotnet tests ... done in 33 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 38 seconds
{
  "args": {
    "bucketName": "minio-go-test-eqw36tefzn0a611f",
    "objectName": "elbj66mhysw3rnnbgqti03562vzhxu",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 4138,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 50 seconds
(10/15) Running minio-js tests ... done in 1 minutes and 2 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 4 seconds
(12/15) Running s3cmd tests ... done in 22 seconds
(13/15) Running s3select tests ... done in 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="2e7c9093-537f-4e96-8c28-a38955cc3f83"></a>
### 10114-7b813ab/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30407
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 32ace75544e8:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 2 minutes and 16 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 22 seconds
(8/15) Running minio-dotnet tests ... done in 52 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 47 seconds
{
  "args": {
    "bucketName": "minio-go-test-6w0t6qroqjy2ocaa",
    "objectName": "ildjsrd5qgtzj3p4uypmzyf6gn9yrr",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 4098,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 1 minutes and 39 seconds
(10/15) Running minio-js tests ... done in 51 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 51 seconds
(12/15) Running s3cmd tests ... done in 37 seconds
(13/15) Running s3select tests ... done in 9 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by poornas** [comment]
> Please also add check to disable all APIs which are not supported on FS mode @poornas - other than GetBucket* calls we should disable all PutBucket* operations.

done

**Review by harshavardhana — APPROVED** [review]

**Review by kannappanr — APPROVED** [review]
LGTM

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#bbeaf260-9e9a-43be-a327-25df888adf3b) |
<a name="bbeaf260-9e9a-43be-a327-25df888adf3b"></a>
### 10114-2a5aca8/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:30556
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fcc97adbfe54:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 3 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 2 minutes and 13 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 50 seconds
(8/15) Running minio-dotnet tests ... done in 49 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 34 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 22 seconds
(11/15) Running minio-js tests ... done in 55 seconds
(12/15) Running minio-py tests ... done in 2 minutes and 34 seconds
(13/15) Running s3cmd tests ... FAILED in 20 seconds
{
  "name": "s3cmd",
  "duration": "8741",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "WARNING: Bucket is not empty. Removing all the objects from it first. This may take some time...\nERROR: S3 error: 404 (Not Found)"
}
(13/15) Running s3select tests ... done in 8 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Pull request merged**